src/sources/socket/udp.rs

```diff
@@ -121,37 +122,52 @@ pub fn udp(
                     emit!(BytesReceived { byte_size, protocol: "udp" });
 
                     let payload = buf.split_to(byte_size);
+                    let truncated = byte_size == max_length + 1;
 
-                    let mut stream = FramedRead::new(payload.as_ref(), decoder.clone());
+                    let mut stream = FramedRead::new(payload.as_ref(), decoder.clone()).peekable();
 
                     while let Some(result) = stream.next().await {
+                        let last = Pin::new(&mut stream).peek().await.is_none();
                         match result {
                             Ok((mut events, _byte_size)) => {
-                                let count = events.len();
-                                emit!(SocketEventsReceived {
-                                    mode: SocketMode::Udp,
-                                    byte_size: events.size_of(),
-                                    count,
-                                });
-
-                                let now = Utc::now();
-
-                                for event in &mut events {
-                                    if let Event::Log(ref mut log) = event {
-                                        log.try_insert(log_schema().source_type_key(), Bytes::from("socket"));
-                                        log.try_insert(log_schema().timestamp_key(), now);
-                                        log.try_insert(host_key.as_str(), address.to_string());
-                                    }
+
+                                if last && truncated {
+                                    // The last event in this payload was truncated, so we want to drop it.
+                                    let _ = events.pop();
+                                    warn!(
+                                        message = "Discarding frame larger than max_length.",
+                                        max_length = max_length,
+                                        internal_log_rate_secs = 30
+                                    );
                                 }
 
-                                tokio::select!{
-                                    result = out.send_batch(events) => {
-                                        if let Err(error) = result {
-                                            emit!(StreamClosedError { error, count });
-                                            return Ok(())
+                                if !events.is_empty() {
```

Comment:
Not a big deal, but I might have done an "early return" pattern here to avoid nesting like:

```rust
if events.is_empty() {
  continue;
}
```

Just a suggestion for the future.